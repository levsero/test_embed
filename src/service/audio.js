const audioStore = {};

export function load(audioFile) {
  const { key, path } = audioFile;
  const audio = new Audio(path);

  audioStore[key] = audio;
}

export function play(key) {
  const audio = audioStore[key];

  if (audio) {
    audio.play();
  }
}
