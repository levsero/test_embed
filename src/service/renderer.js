import { submitTicket } from 'embed/submitTicket/submitTicket';
import { launcher     } from 'embed/launcher/launcher';
import { helpCenter   } from 'embed/helpCenter/helpCenter';
require('imports?_=lodash!lodash');

var embedsMap = {
      'submitTicket': submitTicket,
      'launcher':     launcher,
      'helpCenter':   helpCenter
    },
    initialised = false;

function parseConfig(config) {
  var rendererConfig = _.clone(config, true);

  _.forEach(rendererConfig, function(configItem) {
    configItem.props = _.reduce(configItem.props, function(result, value, key) {
      /* jshint laxbreak: true */
      result[key] = (_.isObject(value))
                  ? function(...args) {
                      args.unshift(value.name);
                      return embedsMap[config[value.name].embed][value.method].apply(null, args);
                    }
                  : value;

      return result;
    }, {});
  });

  return rendererConfig;
}

function init(config) {
  if (!initialised) {
    _.forEach(parseConfig(config), function(configItem, embedName) {
      try {
        embedsMap[configItem.embed].create(embedName, configItem.props);
        embedsMap[configItem.embed].render(embedName);
      } catch (err) {
        // TODO: revisit what this does when error tracking is in place
        console.error(
          'captured error: ',
          {
            embedName: embedName,
            configItem: configItem,
            err: err
          }
        );
      }
    });
  }

  initialised = true;
}

export var renderer = {
  init: init
};
