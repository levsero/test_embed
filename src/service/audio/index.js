import { loadSound, playSound, getSound } from 'service/audio/store';

export const audio = {
  loadSound,
  playSound,
  getSound
};
