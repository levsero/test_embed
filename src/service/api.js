import _ from 'lodash';

import { i18n } from 'service/i18n';
import { mediator } from 'service/mediator';
import { renderer } from 'service/renderer';
import {
  handlePrefillRecieved,
  logout,
  handleOnApiCalled,
  apiClearForm,
  activateRecieved,
  showRecieved,
  hideRecieved,
  legacyShowReceived } from 'src/redux/modules/base';
import { displayArticle, setContextualSuggestionsManually } from 'src/redux/modules/helpCenter';
import { updateSettings } from 'src/redux/modules/settings';
import { chatLogout, sendVisitorPath, endChat, sendMsg } from 'src/redux/modules/chat';
import { getWidgetDisplayInfo } from 'src/redux/modules/base/base-selectors';
import {
  getIsChatting,
  getDepartmentsList,
  getDepartment,
  getNotificationCount,
  getChatStatus } from 'src/redux/modules/chat/chat-selectors';
import {
  API_ON_CHAT_STATUS_NAME,
  API_ON_CLOSE_NAME,
  API_ON_CHAT_CONNECTED_NAME,
  API_ON_CHAT_START_NAME,
  API_ON_CHAT_END_NAME,
  API_ON_CHAT_UNREAD_MESSAGES_NAME,
  API_GET_IS_CHATTING_NAME,
  API_GET_DEPARTMENTS_ALL_NAME,
  API_GET_DEPARTMENTS_DEPARTMENT_NAME,
  API_GET_DISPLAY_NAME } from 'constants/api';
import { CLOSE_BUTTON_CLICKED } from 'src/redux/modules/base/base-action-types';
import {
  CHAT_CONNECTED,
  END_CHAT_REQUEST_SUCCESS,
  NEW_AGENT_MESSAGE_RECEIVED,
  CHAT_STARTED,
  SDK_ACCOUNT_STATUS } from 'src/redux/modules/chat/chat-action-types';

const newAPIPostRenderQueue = [];

const addToPostRenderQueue = (...args) => {
  newAPIPostRenderQueue.push(args);
};
const endChatApi = (reduxStore) => {
  reduxStore.dispatch(endChat());
};
const sendChatMsgApi = (reduxStore, msg) => {
  const message = (_.isString(message)) ? msg : '';

  reduxStore.dispatch(sendMsg(msg));
};
const identifyApi = (reduxStore, user) => {
  mediator.channel.broadcast('.onIdentify', user);
};
const setLocaleApi = (_, locale) => {
  console.log('here');
  i18n.setLocale(locale, true);
  mediator.channel.broadcast('.onSetLocale', locale);
};
const updateSettingsApi = (reduxStore, newSettings) => {
  reduxStore.dispatch(updateSettings(newSettings));
};
const logoutApi = (reduxStore) => {
  reduxStore.dispatch(logout());
  mediator.channel.broadcast('.logout');
  reduxStore.dispatch(chatLogout());
};
const setHelpCenterSuggestionsApi = (reduxStore, options) => {
  const onDone = () => mediator.channel.broadcast('.setHelpCenterSuggestions');

  reduxStore.dispatch(setContextualSuggestionsManually(options, onDone));
};
const prefill = (reduxStore, payload) => {
  reduxStore.dispatch(handlePrefillRecieved(payload));
};
const hideApi = (reduxStore) => {
  reduxStore.dispatch(hideRecieved());
};
const showApi = (reduxStore) => {
  reduxStore.dispatch(showRecieved());
};
const updatePathApi = (reduxStore, page = {}) => {
  reduxStore.dispatch(sendVisitorPath(page));
};
const getWidgetChatApiObj = () => {
  return {
    end: endChatApi,
    send: sendChatMsgApi
  };
};
const onApiObj = () => {
  const chatEventMap = {
    [API_ON_CHAT_CONNECTED_NAME]: { actionType: CHAT_CONNECTED },
    [API_ON_CHAT_END_NAME]: { actionType: END_CHAT_REQUEST_SUCCESS },
    [API_ON_CHAT_START_NAME]: { actionType: CHAT_STARTED },
    [API_ON_CHAT_STATUS_NAME]: {
      actionType: SDK_ACCOUNT_STATUS,
      selectors: [getChatStatus]
    },
    [API_ON_CHAT_UNREAD_MESSAGES_NAME]: {
      actionType: NEW_AGENT_MESSAGE_RECEIVED,
      selectors: [getNotificationCount]
    }
  };
  const baseEventMap = {
    [API_ON_CLOSE_NAME]: { actionType: CLOSE_BUTTON_CLICKED }
  };
  const eventDispatchWrapperFn = (actionType, selectors = []) => {
    return (reduxStore, callback) => {
      if (_.isFunction(callback)) {
        reduxStore.dispatch(handleOnApiCalled(actionType, selectors, callback));
      }
    };
  };
  const eventApiReducerFn = (eventMap) => {
    return _.reduce(eventMap, (apiObj, eventObj, eventName) => {
      const { actionType, selectors } = eventObj;

      apiObj[eventName] = eventDispatchWrapperFn(actionType, selectors);

      return apiObj;
    }, {});
  };

  return {
    'chat': eventApiReducerFn(chatEventMap),
    ...eventApiReducerFn(baseEventMap)
  };
};
const getApiObj = () => {
  return {
    chat: {
      [API_GET_IS_CHATTING_NAME]: (reduxStore, ...args) => getIsChatting(reduxStore.getState(), ...args),
      [API_GET_DEPARTMENTS_ALL_NAME]: (reduxStore, ...args) => getDepartmentsList(reduxStore.getState(), ...args),
      [API_GET_DEPARTMENTS_DEPARTMENT_NAME]: (reduxStore, ...args) => getDepartment(reduxStore.getState(), ...args),
    },
    [API_GET_DISPLAY_NAME]: (reduxStore, ...args) => getWidgetDisplayInfo(reduxStore.getState(), ...args)
  };
};
const getApiPostRenderQueue = () => {
  const postRenderCallback = (_, ...args) => addToPostRenderQueue(['webWidget:get', ...args]);

  return {
    [API_GET_IS_CHATTING_NAME]: postRenderCallback,
    [API_GET_DEPARTMENTS_ALL_NAME]: postRenderCallback,
    [API_GET_DEPARTMENTS_DEPARTMENT_NAME]: postRenderCallback,
    [API_GET_DISPLAY_NAME]: postRenderCallback
  };
};

const newApiStructurePostRender = {
  webWidget: {
    hide: hideApi,
    show: showApi,
    setLocale: setLocaleApi,
    identify: identifyApi,
    updateSettings: updateSettingsApi,
    logout: logoutApi,
    setSuggestions: setHelpCenterSuggestionsApi,
    updatePath: updatePathApi,
    clear: clearFormState,
    prefill: prefill,
    chat: getWidgetChatApiObj(),
    on: onApiObj(),
    get: getApiObj()
  }
};
const newApiStructurePreRender = {
  webWidget: {
    hide: hideApi,
    show: (_, ...args) => addToPostRenderQueue(['webWidget', 'show', ...args]),
    setLocale: (_, locale) => i18n.setLocale(locale),
    identify: (_, ...args) => addToPostRenderQueue(['webWidget', 'identify', ...args]),
    updateSettings: (_, ...args) => addToPostRenderQueue(['webWidget', 'updateSettings', ...args]),
    logout: (_, ...args) => addToPostRenderQueue(['webWidget', 'logout', ...args]),
    setSuggestions: (_, ...args) => {
      addToPostRenderQueue(['webWidget', 'setSuggestions', ...args]);
    },
    updatePath: (_, ...args) => addToPostRenderQueue(['webWidget', 'updatePath', ...args]),
    clear: (reduxStore) => clearFormState(reduxStore),
    prefill: prefill,
    chat: getWidgetChatApiObj(),
    on: onApiObj(),
    get: getApiPostRenderQueue(),
  }
};

function clearFormState(reduxStore) {
  reduxStore.dispatch(apiClearForm());
}

const handleNewApi = (apiStructure, reduxStore, args) => {
  const getApiFunction = (methodAccessors) => {
    const keys = _.flatten(methodAccessors.map((accessor => {
      return accessor.split(':');
    }))).join('.');

    return _.get(apiStructure, keys, () => {});
  };
  const params = Array.from(args);

  /*
   Assume the first two arguments provided by the user will be method accessor params.
   Any subsequent parameters provided will be additional data for whatever api they are calling (eg. callbacks).
  */
  const methodAccessorParams = params.slice(0, 2);
  const apiMethodParams = params.slice(2);
  const apiFunction = getApiFunction(methodAccessorParams);

  return apiFunction(reduxStore, ...apiMethodParams);
};

function handleQueue(reduxStore, queue) {
  const logApiError = (api, e = {}) => {
    const err = new Error([
      'An error occurred in your use of the Zendesk Widget API:',
      api,
      'Check out the Developer API docs to make sure you\'re using it correctly',
      'https://developer.zendesk.com/embeddables/docs/widget/api',
      e.stack
    ].join('\n\n'));

    err.special = true;
    throw err;
  };

  _.forEach(queue, (method) => {
    if (method[0].locale) {
      // Backwards compat with zE({locale: 'zh-CN'}) calls
      i18n.setLocale(method[0].locale);
    } else if (_.isFunction(method[0])) {
      // Old API
      try {
        method[0]();
      } catch (e) {
        logApiError(method[0], e);
      }
    } else if (_.includes(method[0], 'webWidget')) {
      // New API
      try {
        handleNewApi(newApiStructurePreRender, reduxStore, method);
      } catch (e) {
        logApiError(`"${method[0]} ${method[1]}"`, e);
      }
    } else {
      logApiError(method);
    }
  });
}

function handlePostRenderQueue(win, postRenderQueue, reduxStore) {
  _.forEach(postRenderQueue, (method) => {
    win.zE[method[0]](...method[1]);
  });

  _.forEach(newAPIPostRenderQueue, (item) => {
    handleNewApi(newApiStructurePostRender, reduxStore, ...item);
  });

  renderer.postRenderCallbacks();
}

function setupWidgetQueue(win, postRenderQueue, reduxStore) {
  let devApi;

  // no "fat arrow" because it binds `this` to the scoped environment and does not allow it to be re-set with .bind()
  const postRenderQueueCallback = function(...args) {
    // "this" is bound to the method name
    postRenderQueue.push([this, args]);
  };
  const publicApi = {
    version: __EMBEDDABLE_VERSION__,
    setLocale: i18n.setLocale,
    hide: hideApi,
    show: postRenderQueueCallback.bind('show'),
    setHelpCenterSuggestions: postRenderQueueCallback.bind('setHelpCenterSuggestions'),
    identify: postRenderQueueCallback.bind('identify'),
    logout: postRenderQueueCallback.bind('logout'),
    activate: postRenderQueueCallback.bind('activate'),
    configureIPMWidget: postRenderQueueCallback.bind('configureIPMWidget'),
    showIPMArticle: postRenderQueueCallback.bind('showIPMArticle'),
    hideIPMWidget: postRenderQueueCallback.bind('hideIPMWidget'),
    activateIpm: () => {} // no-op until rest of connect code is removed
  };

  if (__DEV__) {
    devApi = {
      devRender: (config) => {
        if (config.ipmAllowed) {
          setupIPMApi(win, reduxStore, config);
        }
        renderer.init(config, reduxStore);
      }
    };
  }

  if (win.zE === win.zEmbed) {
    win.zE = win.zEmbed = (...args) => {
      if (_.isFunction(args[0])) {
        args[0]();
      } else {
        return handleNewApi(newApiStructurePostRender, reduxStore, args);
      }
    };
  } else {
    win.zEmbed = (...args) => {
      if (_.isFunction(args[0])) {
        args[0]();
      } else {
        return handleNewApi(newApiStructurePostRender, reduxStore, args);
      }
    };
  }

  return {
    publicApi,
    devApi
  };
}

function setupZopimQueue(win) {
  let $zopim = () => {};

  // To enable $zopim api calls to work we need to define the queue callback.
  // When we inject the snippet we remove the queue method and just inject
  // the script tag.
  if (!win.$zopim) {
    $zopim = win.$zopim = (callback) => {
      $zopim._.push(callback);
    };

    $zopim.set = (callback) => {
      $zopim.set._.push(callback);
    };
    $zopim._ = [];
    $zopim.set._ = [];
  }
}

function setupIPMApi(win, reduxStore, embeddableConfig = {}) {
  const existingConfig = !_.isEmpty(embeddableConfig.embeds);
  const prefix = existingConfig ? '' : 'ipm.';

  win.zE.configureIPMWidget = (config) => {
    if (!existingConfig) {
      renderer.initIPM(config, embeddableConfig, reduxStore);
    }
  };
  win.zE.showIPMArticle = (articleId) => {
    reduxStore.dispatch(displayArticle(articleId));
  };
  win.zE.showIPMWidget = () => {
    mediator.channel.broadcast(`${prefix}webWidget.show`);
  };
  win.zE.hideIPMWidget = () => {
    mediator.channel.broadcast(`${prefix}webWidget.hide`);
  };
}

function setupWidgetApi(win, reduxStore) {
  win.zE.identify = (user) => {
    identifyApi(reduxStore, user);

    if (!user || (!user.email || !user.name)) return;

    const prefillUser = {
      name: { value: user.name },
      email: { value: user.email }
    };

    prefill(reduxStore, prefillUser);
  };
  win.zE.logout = () => logoutApi(reduxStore);
  win.zE.setHelpCenterSuggestions = (options) => setHelpCenterSuggestionsApi(reduxStore, options);
  win.zE.activate = (options) => {
    mediator.channel.broadcast('.activate', options);
    reduxStore.dispatch(activateRecieved());
  };
  win.zE.activateIpm = () => {}; // no-op until rest of connect code is removed
  win.zE.hide = () => hideApi(reduxStore);
  win.zE.show = () => reduxStore.dispatch(legacyShowReceived());
  win.zE.setLocale = (locale) => setLocaleApi(null, locale);
}

export const api = {
  handleQueue,
  handlePostRenderQueue,
  setupWidgetQueue,
  setupZopimQueue,
  setupWidgetApi,
  setupIPMApi
};
