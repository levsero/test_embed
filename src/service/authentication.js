import { mediator } from 'service/mediator';
import { transport } from 'service/transport';

function init() {
  mediator.channel.subscribe('authentication.authenticate', authenticate);
}

function authenticate(webToken) {
  const payload = {
    method: 'POST',
    path: '/embeddable/authenticate',
    params: webToken,
    callbacks: {
      done: function(res) {
        // TODO: logic depending on the status we get
        return res;
      },
      fail: function(err) {
        return err;
      }
    }
  };

  transport.send(payload);
}

export const authentication = {
  init: init,
  authenticate: authenticate
};
