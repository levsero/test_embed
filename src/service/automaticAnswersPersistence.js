import { store } from 'service/persistence';
import { getURLParameterByName } from 'utility/pages';

const automaticAnswersLocalStorageKey = 'automaticAnswers';
const automaticAnswersJwtUrlParameter = 'auth_token';

function setLocalStorage(authToken) {
  const currentTime = Date.now();
  const oneDayInMs = 86400000;

  store.set(
      automaticAnswersLocalStorageKey,
      { authToken: authToken, expiry: currentTime + oneDayInMs }
  );
}

function getFromLocalStorage() {
  return store.get(automaticAnswersLocalStorageKey);
}

function authTokenFromStore() {
  const automaticAnswersConfig = getFromLocalStorage();

  if (!automaticAnswersConfig) return null;
  if (Date.now() > automaticAnswersConfig.expiry) {
    store.remove(automaticAnswersLocalStorageKey);
    return null;
  }

  return automaticAnswersConfig.authToken;
}

function getContext() {
  const authTokenFromUrl = getURLParameterByName(automaticAnswersJwtUrlParameter);

  if (authTokenFromUrl) {
    setLocalStorage(authTokenFromUrl);
    return authTokenFromUrl;
  } else {
    return authTokenFromStore();
  }
}

export const automaticAnswersPersistence = {
  getContext
};
