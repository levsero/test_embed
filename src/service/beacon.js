import { win,
         document,
         navigator } from 'util/globals';
import { transport } from 'service/transport';
import { identity  } from 'service/identity';
import { store     } from 'service/persistence';
import { parseUrl  } from 'util/utils';

function init() {
  var now = Date.now();
  store.set('currentTime', now, true);
  return this;
}

function send() {
  var now = Date.now(),
      referrer = parseUrl(document.referrer),
      previousTime = store.get('currentTime', true) || 0,
      url = win.location.origin,
      timeOnLastPage = function () {
        return referrer.origin === url && previousTime ? (now - previousTime) : 0;
      },

      params = {
        'url': win.location.href,
        'buid': identity.getBuid(),
        'user_agent': navigator.userAgent,
        'referrer': referrer.href,
        'time': timeOnLastPage(),
        'metrics': ['beacon']
      };

  var payload = {
    method: 'POST',
    path: '/api/blips',
    params: params,
    callbacks: {
      done: function() {},
      fail: function() {}
    }
  };

  transport.send(payload);
}

export var beacon = {
  init: init,
  send: send
};
