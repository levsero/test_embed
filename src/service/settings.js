import _ from 'lodash';

import { win } from 'utility/globals';
import { objectDifference } from 'utility/utils';
import { updateSettings } from 'src/redux/modules/settings';
import { mediator } from 'service/mediator';

const optionWhitelist = {
  webWidget: [
    'authenticate',
    'authenticate.support',
    'authenticate.chat',
    'contactOptions.enabled',
    'contactOptions.contactButton',
    'contactOptions.chatLabelOnline',
    'contactOptions.chatLabelOffline',
    'contactOptions.contactFormLabel',
    'chat.concierge.avatarPath',
    'chat.departments.enabled',
    'chat.suppress',
    'chat.departments.select',
    'chat.notifications.mobile.disable',
    'chat.prechatForm.departmentLabel',
    'chat.title',
    'chat.prechatForm.greeting',
    'chat.offlineForm.greeting',
    'chat.concierge.title',
    'chat.concierge.name',
    'chat.hideWhenOffline',
    'chat.tags',
    'color.theme',
    'color.button',
    'color.header',
    'color.launcher',
    'color.launcherText',
    'color.articleLinks',
    'color.resultLists',
    'contactForm.attachments',
    'contactForm.fields',
    'contactForm.selectTicketForm',
    'contactForm.subject',
    'contactForm.suppress',
    'contactForm.tags',
    'contactForm.ticketForms',
    'contactForm.title',
    'helpCenter.chatButton',
    'helpCenter.filter',
    'helpCenter.localeFallbacks',
    'helpCenter.messageButton',
    'helpCenter.originalArticleButton',
    'helpCenter.searchPlaceholder',
    'helpCenter.suppress',
    'helpCenter.title',
    'launcher.chatLabel',
    'launcher.label',
    'launcher.badge.layout',
    'launcher.badge.image',
    'launcher.badge.label',
    'offset.horizontal',
    'offset.vertical',
    'offset.mobile.horizontal',
    'offset.mobile.vertical',
    'position.horizontal',
    'position.vertical',
    'talk.nickname',
    'talk.suppress',
    'zIndex'
  ]
};
const customizationsWhitelist = [
  'helpCenter.localeFallbacks'
];
const webWidgetStoreDefaults = {
  contactForm: {
    attachments: true,
    fields: [],
    subject: false,
    suppress: false,
    tags: [],
    ticketForms: []
  },
  contactOptions: { enabled: false },
  chat: {
    concierge: {
      avatarPath: null
    },
    departments: {
      enabled: null,
      select: ''
    },
    suppress: false,
    notifications: {
      mobile: {
        disable: false
      }
    },
    tags: []
  },
  launcher: { },
  margin: 8,
  offset: {
    horizontal: 0,
    vertical: 0,
    mobile: {
      horizontal: 0,
      vertical: 0
    }
  },
  talk: {
    suppress: false
  },
  viaId: 48,
  zIndex: 999999
};
const baseDefaults = {
  errorReporting: true,
  analytics: true
};
let settingsStore = {};
let webWidgetStore = {};
let webWidgetCustomizations = false;

const initStore = (settings, options, defaults) => {
  const reduceFn = (res, val) => {
    if (_.has(settings, val)) {
      _.set(res, val, _.get(settings, val, null));
    }
    return res;
  };

  return _.chain(options)
    .reduce(reduceFn, {})
    .defaultsDeep(defaults)
    .value();
};

function init(reduxStore = { dispatch: () => {} }) {
  settingsStore = _.assign({}, baseDefaults, win.zESettings);

  // for backwards compatibility with authenticate.
  if (settingsStore.authenticate) {
    if (!settingsStore.webWidget) {
      settingsStore.webWidget = {};
    }
    settingsStore.webWidget.authenticate = settingsStore.authenticate;
  }

  webWidgetStore = initStore(settingsStore.webWidget, optionWhitelist.webWidget, webWidgetStoreDefaults);

  reduxStore.dispatch(updateSettings({
    webWidget: {
      ...webWidgetStore,
      ...settingsStore
    }
  }));
}

function get(path) {
  // TODO: Remove this check when web widget customizations are out of beta.
  if (customizationsWhitelist.indexOf(path) > -1 &&
      !webWidgetCustomizations) {
    return _.get(webWidgetStoreDefaults, path, null);
  }

  return _.get(webWidgetStore, path, null);
}

function getTranslations() {
  const translations = {
    contactFormSelectTicketForm: webWidgetStore.contactForm.selectTicketForm,
    contactFormTitle: webWidgetStore.contactForm.title,
    helpCenterChatButton: webWidgetStore.helpCenter.chatButton,
    helpCenterMessageButton: webWidgetStore.helpCenter.messageButton,
    helpCenterContactButton: webWidgetStore.contactOptions.contactButton,
    helpCenterSearchPlaceholder: webWidgetStore.helpCenter.searchPlaceholder,
    helpCenterTitle: webWidgetStore.helpCenter.title,
    launcherChatLabel: webWidgetStore.launcher.chatLabel,
    launcherLabel: webWidgetStore.launcher.label,
    contactOptionsChatLabelOnline: webWidgetStore.contactOptions.chatLabelOnline,
    contactOptionsChatLabelOffline: webWidgetStore.contactOptions.chatLabelOffline,
    contactOptionsContactFormLabel: webWidgetStore.contactOptions.contactFormLabel
  };

  return _.omitBy(translations, _.isUndefined);
}

function getTrackSettings() {
  const blacklist = ['margin', 'viaId'];
  const userSettings = _.omit(webWidgetStore, blacklist);
  const defaults = _.omit(webWidgetStoreDefaults, blacklist);
  const widgetSettings = objectDifference(userSettings, defaults);

  if (widgetSettings.authenticate) {
    widgetSettings.authenticate = true;
  }

  return _.omitBy({
    webWidget: widgetSettings
  }, _.isEmpty);
}

function getSupportAuthSettings() {
  let authSetting = get('authenticate.support');

  if (authSetting && authSetting.jwt) {
    return authSetting;
  }

  // If webWidget.authenticate.support setting is not valid.
  // Fallback to original webWidget.authenticate setting
  authSetting = get('authenticate');

  return (authSetting && authSetting.jwt) ? authSetting : null;
}

function getChatAuthSettings() {
  const authSetting = get('authenticate.chat');

  return (authSetting && authSetting.jwtFn) ? authSetting : null;
}

function getErrorReportingEnabled() {
  return Boolean(settingsStore.errorReporting);
}

function enableCustomizations() {
  webWidgetCustomizations = true;
}

function updateSettingsLegacy(newSettings, callback=() => {}) {
  _.merge(webWidgetStore, newSettings);
  callback();
  mediator.channel.broadcast('.onUpdateSettings');
}

export const settings = {
  init,
  get,
  getTranslations,
  getTrackSettings,
  getSupportAuthSettings,
  getChatAuthSettings,
  getErrorReportingEnabled,
  enableCustomizations,
  updateSettingsLegacy
};
