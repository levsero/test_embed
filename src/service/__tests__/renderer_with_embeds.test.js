import 'utility/i18nTestHelper';
import { renderer } from '../renderer';
import _ from 'lodash';
import { settings } from 'service/settings';
import { createStore } from 'redux';
import reducer from 'src/redux/modules/reducer';

jest.mock('service/settings');
jest.mock('src/redux/modules/base');

const store = createStore(reducer);
let mockSettings;

store.dispatch = jest.fn();

beforeEach(() => {
  mockSettings = {
    contactOptions: { enabled: false },
    offset: { vertical: 20, horizontal: 30 }
  };
  settings.get = (value) => _.get(mockSettings, value, null);
});

test('renders expected embeds from config', () => {
  const configJSON = {
    embeds: {
      'helpCenterForm': {
        'embed': 'helpCenter',
        'props': {}
      },
      'launcher': {
        'embed': 'launcher',
        'props': {
          'position': 'right'
        }
      },
      'ticketSubmissionForm': {
        'embed': 'submitTicket'
      },
      'zopimChat': {
        'embed': 'chat',
        'props': {
          'zopimId': '2EkTn0An31opxOLXuGgRCy5nPnSNmpe6',
          'position': 'br'
        }
      }
    }
  };

  renderer.init(configJSON, store);

  expect(document.body.innerHTML)
    .toMatchSnapshot();
});
