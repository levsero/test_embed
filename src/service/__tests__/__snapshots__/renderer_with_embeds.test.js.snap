// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders expected embeds from config 1`] = `
"<div><iframe title=\\"Opens a widget where you can find more information\\" style=\\"background: transparent; z-index: 999998; transform: translateZ(0); position: fixed; opacity: 0; width: 347px; height: 50px; max-height: 551px; min-height: 50px; margin: 10px 20px 10px 20px; top: -9999px; visibility: hidden;\\" id=\\"launcher\\" tabindex=\\"-1\\" class=\\"zEWidget-launcher \\"></iframe></div><script>
    (function(d,s){var z=$zopim,$=z.s= d.createElement(s),e=d.getElementsByTagName(s)[0];
    $.async=!0;$.setAttribute('charset','utf-8');
    $.src='https://v2.zopim.com/w?2EkTn0An31opxOLXuGgRCy5nPnSNmpe6';
    z.t=+new Date;$. type='text/javascript';e.parentNode.insertBefore($,e)})(document,'script');
  </script><div><iframe title=\\"Find more information here\\" style=\\"background: transparent; z-index: 999999; transform: translateZ(0); position: fixed; opacity: 0; width: 357px; height: 100%; max-height: 565px; min-height: 150px; top: -9999px; visibility: hidden;\\" id=\\"webWidget\\" tabindex=\\"-1\\" class=\\"zEWidget-webWidget \\"></iframe></div>"
`;
